KAFKA-4306: Shutdown distributed herder with a timeout.

kkonstantine opened the pull request:
Resolves

KAFKA-4306: Connect workers won't shut down if brokers are not available
KAFKA-4154: Kafka Connect fails to shutdown if it has not completed startup

kkonstantine committed:
KAFKA-4306: Shutdown distributed herder with a timeout.

Resolves

KAFKA-4306: Connect workers won't shut down if brokers are not available
KAFKA-4154: Kafka Connect fails to shutdown if it has not completed startup

shikhar commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -372,20 +379,17 @@ public void stop() {
+        herderExecutor.shutdown();
         try {
-            if (!forwardRequestExecutor.awaitTermination(10000, TimeUnit.MILLISECONDS))
+            if (!forwardRequestExecutor.awaitTermination(workerTasksShutdownTimeoutMs, TimeUnit.MILLISECONDS))
```
This should probably refer to `herderExecutor` :)

kkonstantine replied:
yes!

kkonstantine committed:
KAFKA-3008: Parallel start and stop of connectors and tasks in Connect

Author: Konstantine Karantasis <konstantine@confluent.io>
Author: Konstantine Karantasis <k.karantasis@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Shikhar Bhushan <shikhar@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1788 from kkonstantine/KAFKA-3008-Parallel-start-and-stop-of-connectors-and-tasks

shikhar committed:
KAFKA-4161: KIP-89: Allow sink connectors to decouple flush and offset commit

Author: Shikhar Bhushan <shikhar@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2139 from shikhar/kafka-4161-deux

resetius committed:
KAFKA-4399; Fix deadlock between cleanupGroupMetadata and offset commit

Author: Alexey Ozeritsky <aozeritsky@yandex-team.ru>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jason Gustafson <jason@confluent.io>

Closes #2125 from resetius/KAFKA-4399

ijuma committed:
KAFKA-2247; Merge kafka.utils.Time and kafka.common.utils.Time

Also:
* Make all implementations of `Time` thread-safe as they are accessed from multiple threads in some cases.
* Change default implementation of `MockTime` to use two separate variables for `nanoTime` and `currentTimeMillis` as they have different `origins`.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Shikhar Bhushan <shikhar@confluent.io>, Jason Gustafson <jason@confluent.io>, Eno Thereska <eno.thereska@gmail.com>, Damian Guy <damian.guy@gmail.com>

Closes #2095 from ijuma/kafka-2247-consolidate-time-interfaces

enothereska committed:
2198

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2198 from enothereska/hotfix-stream-states

hachikuji committed:
MINOR: Improvements in group metadata cleanup and test coverage

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma, Guozhang Wang

Closes #2202 from hachikuji/group-expiration-cleanup

ijuma committed:
KAFKA-4483; Fix NPE in `Log` constructor if log level is INFO or finer

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2207 from ijuma/kafka-4483-npe-in-log-constructor

ataraxer committed:
KAFKA-4205; KafkaApis: fix NPE caused by conversion to array

NPE was caused by `log.logSegments.toArray` resulting in array containing `null` values. The exact reason still remains somewhat a mystery to me, but it seems that the culprit is `JavaConverters` in combination with concurrent data structure access.

Here's a simple code example to prove that:
```scala
import java.util.concurrent.ConcurrentSkipListMap
// Same as `JavaConversions`, but allows explicit conversions via `asScala`/`asJava` methods.
import scala.collection.JavaConverters._

case object Value
val m = new ConcurrentSkipListMap[Int, Value.type]
new Thread { override def run() = { while (true) m.put(9000, Value) } }.start()
new Thread { override def run() = { while (true) m.remove(9000) } }.start()
new Thread { override def run() = { while (true) { println(m.values.asScala.toArray.headOption) } } }.start()
```

Running the example will occasionally print `Some(null)` indicating that there's something shady going on during `toArray` conversion.

`null`s magically disappear by making the following change:
```diff
- println(m.values.asScala.toArray.headOption)
+ println(m.values.asScala.toSeq.headOption)
```

Author: Anton Karamanov <ataraxer@yandex-team.ru>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Guozhang Wang <wangguoz@gmail.com>

Closes #2204 from ataraxer/KAFKA-4205

hachikuji committed:
KAFKA-3994: Fix deadlock in Watchers by calling tryComplete without any locks

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma, Jun Rao, Jiangjie Qin, Guozhang Wang

Closes #2195 from hachikuji/KAFKA-3994-linked-queue

kkonstantine committed:
HOTFIX: Fix bug in readToLogEnd in KafkaBasedLog.

Author: Konstantine Karantasis <konstantine@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2211 from kkonstantine/HOTFIX-Correctly-read-to-end-of-offsets-log-in-Connect-KafkaBasedLog

ewencp commented:
@kkonstantine I think something got messed up with the branch and some unrelated changes are getting included.

ewencp commented:
Oh, nm, had this stale PR open instead of the one against the Kafka repo.

Commit:
KAFKA-4472; offsetRetentionMs miscalculated in GroupCoordinator

Fix possible integer overflow.

Author: Kim Christensen <kich@mvno.dk>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2200 from kichristensen/MiscalculatedOffsetRetention

enothereska committed:
HOTFIX: Temporary suspension of 2 tests

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2199 from enothereska/hotfix-streams-test-reset-ignore
